### ISSUE TITLE
PPS: no geometry specification in reco cff files

### ISSUE THREAD

**Issue by jan-kaspar** [cms-sw/cmssw#31360]
It has been recommended to PPS to remove geometry specifications from reco cff files, see e.g.
  * https://github.com/cms-sw/cmssw/pull/31240#issuecomment-686823112
  * https://github.com/cms-sw/cmssw/pull/31240#issuecomment-687014268

**Comment by jan-kaspar** [comment]
FYI @makortel @davidlange6

**Comment by jan-kaspar** [comment]
@davidlange6 Who would be the right contacts from the geometry system to coordinate this work with?

**Comment by makortel** [comment]
assign geometry

**Comment by makortel** [comment]
> Who would be the right contacts from the geometry system to coordinate this work with?

@cvuosalo (and @ianna )

**Comment by cvuosalo** [comment]
@jan-kaspar I think the correct place for PPS reco geometry might be `Configuration/Geometry/python/GeometryRecoDB_cff.py`. XML and simulation geometry should be checked to be sure they include all needed PPS components. Additional cff files in `Configuration/Geometry` like `GeometryReco_cff.py` may need additions for PPS.

**Comment by jan-kaspar** [comment]
Thanks @cvuosalo ! What is the cff file for the "XML geometry" that you mention? What is the use of the "GeometryReco_cff.py"?

**Comment by jan-kaspar** [comment]
If you agreed, I would schedule this fix after the other ongoing geometry-related modifications by @ghugo83 and @forthommel - not to do to many things in parallel.

**Comment by cvuosalo** [comment]
@jan-kaspar `Configuration/​Geometry/​python` contains all the scenarios. The scenarios that use the DB have "DB" in the name. If the scenario doesn't have "DB" in the name, it is derived from XML, so `GeometryReco_cff.py` is for (Run 2) Reco geometry from XML.
What time periods does `geometryRPFromDB_cfi.py` apply to? Caution must be exercised with changes to files used in existing workflows like `GeometryRecoDB_cff.py`

**Comment by jan-kaspar** [comment]
> @jan-kaspar `Configuration/​Geometry/​python` contains all the scenarios. The scenarios that use the DB have "DB" in the name. If the scenario doesn't have "DB" in the name, it is derived from XML, so `GeometryReco_cff.py` is for (Run 2) Reco geometry from XML.

Thanks for clarification.

> What time periods does `geometryRPFromDB_cfi.py` apply to?

2016, 2017 and 2018 (the full life of PPS)

> Caution must be exercised with changes to files used in existing workflows like `GeometryRecoDB_cff.py`

I guess that the usual matrix tests are a good way to check the update. Would there be anything else to verify?

**Comment by cvuosalo** [comment]
@jan-kaspar It should be checked that every config that currently uses `recoCTPPS_cff.py` will still get `geometryRPFromDB_cfi.py` after the latter is moved.

**Comment by cvuosalo** [comment]
Since `geometryRPFromDB_cfi.py` is only used for Run 2, I wonder if we really want to make any change related to it. What are the PPS plans for changing Run 2?

**Comment by jan-kaspar** [comment]
> Since `geometryRPFromDB_cfi.py` is only used for Run 2, I wonder if we really want to make any change related to it. What are the PPS plans for changing Run 2?

The file is used for Run2 at the moment, but it would be used for Run3 once would get there.

Currently PPS has no plans for updating Run2 geometry (beyond the DD4Hep efforts).

Personally, I feel no urgent need to change anything. However, it has bee argued by @davidlange6 @makortel that the structure that we have doesn't follow CMS standards.

**Comment by cvuosalo** [comment]
@jan-kaspar Let's leave Run 2 alone. When PPS starts using these files for Run 3 (if ever), then they can be revised as we discussed. In the latest PR #31383, Gabrielle has been careful not to alter existing workflows, so I suspect that Run 3 development will create new versions of these files and leave the old ones unchanged.

**Comment by cvuosalo** [comment]
-1

**Comment by cvuosalo** [comment]
@davidlange6 @makortel In #31240, files used by Run 2 were modified that caused problems with existing workflows. That led to your comments [1] [2] about the incorrect structure of `recoCTPPS_cff.py`. However, it was incorrect to modify the Run 2 files in the first place. PR #31383 replaces #31240, and the new PR carefully avoids changing existing workflows. The issue you raised has been long-standing in Run 2, and I don't think we want to tamper with it now. Going forward for Run 3 development, we will address this issue in new code, but let's leave Run 2 unchanged.

[1]: https://github.com/cms-sw/cmssw/pull/31240#issuecomment-686823112
[2]: https://github.com/cms-sw/cmssw/pull/31240#issuecomment-687014268

**Comment by cvuosalo** [comment]
@jan-kaspar I think you can close this issue, or keep it open as a reminder of how to implement Run 3 PPS Reco geometry.

**Comment by makortel** [comment]
@cvuosalo Do you specifically mean that the geometry import
https://github.com/cms-sw/cmssw/blob/04068a6ee2b79646b2005758083db57a92840ac1/RecoPPS/Configuration/python/recoCTPPS_cff.py#L12
would be left as it is? Are you sure the two ESProducers
https://github.com/cms-sw/cmssw/blob/04068a6ee2b79646b2005758083db57a92840ac1/Geometry/VeryForwardGeometry/python/geometryRPFromDB_cfi.py#L3-L12
(from Run2) will not interfere with anything for Run3 (and beyond)? (I'm genuinely asking, I have no idea)

**Comment by cvuosalo** [comment]
@makortel All the code you are pointing to is old DD code. In Run 3, its purpose is for validation comparisons. I would guess that it works OK. `geometryRPFromDB_cfi.py` has been migrated to DD4hep in #31383, and we hope that will be the production version for Run 3, so the old code you point to won't be affecting future development. When the PPS developers revise `recoCTPPS_cff.py` for Run 3, it will be re-structured as you and David have suggested.

**Comment by jan-kaspar** [comment]
> ...  When the PPS developers revise `recoCTPPS_cff.py` for Run 3, it will be re-structured as you and David have suggested.

If possible, we'd like to maintain Run2-compatibility. One of the reasons is that running over Run2 data provides a very useful SW-integrity test. But I guess this can be achieved e.g. with era-dependent configuration.

**Comment by jan-kaspar** [comment]
@makortel @davidlange6 Do you agree with the suggestion to close this issue?

**Comment by davidlange6** [comment]
I don’t, no.

**Comment by jan-kaspar** [comment]
I possibly can't see the full chain of implications, but it seems to me that we're talking about a rather small and technical change: essentially move the include of "geometryRPFromDB_cfi" from a PPS reco cff to a global geometry cff. Is it really that difficult/dangerous?

**Comment by cvuosalo** [comment]
@jan-kaspar From my point of view, PPS DD4hep migration has been encountering numerous issues, which we are in the process of resolving while under significant time pressure. This issue concerns the structure of Run 2, which is working OK right now, even if it isn't structured exactly as we would like. Yes, the fix may be fairly simple, but since it deals with existing workflows, it will have to be checked carefully. I would rather defer Run 2 re-structuring until after the urgent PPS development for Run 3 is completed.

**Comment by jan-kaspar** [comment]
> ... I would rather defer Run 2 re-structuring until after the urgent PPS development for Run 3 is completed.

I fully agree! Would deferring this issue be OK for @davidlange6 and @makortel ?
